engine/consensus/compliance/engine.go

```diff
@@ -44,39 +45,45 @@ const defaultTimeoutObjectsQueueCapacity = 1000
 // and forwards outbound messages to the networking layer.
 // `compliance.Core` implements the actual compliance logic.
 type Engine struct {
-	unit                       *engine.Unit
-	lm                         *lifecycle.LifecycleManager
-	log                        zerolog.Logger
-	mempool                    module.MempoolMetrics
-	metrics                    module.EngineMetrics
-	me                         module.Local
-	headers                    storage.Headers
-	payloads                   storage.Payloads
-	tracer                     module.Tracer
-	state                      protocol.State
-	prov                       network.Engine
-	core                       *Core
-	pendingBlocks              engine.MessageStore
-	pendingRangeResponses      engine.MessageStore
-	pendingVotes               engine.MessageStore
-	pendingTimeouts            engine.MessageStore
-	messageHandler             *engine.MessageHandler
+	log      zerolog.Logger
+	mempool  module.MempoolMetrics
+	metrics  module.EngineMetrics
+	me       module.Local
+	headers  storage.Headers
+	payloads storage.Payloads
+	tracer   module.Tracer
+	state    protocol.State
+	prov     consensus.ProposalProvider
+	core     *Core
+	// queues for inbound messsages
+	pendingBlocks         engine.MessageStore
+	pendingRangeResponses engine.MessageStore
+	// TODO remove pendingVotes and pendingTimeouts - we will pass these directly to the Aggregator
+	pendingVotes    engine.MessageStore
+	pendingTimeouts engine.MessageStore
+	messageHandler  *engine.MessageHandler
+	// tracking finalized view
 	finalizedView              counters.StrictMonotonousCounter
 	finalizationEventsNotifier engine.Notifier
 	con                        network.Conduit
-	stopHotstuff               context.CancelFunc
+
+	cm *component.ComponentManager
+	component.Component
 }
 
+var _ network.MessageProcessor = (*Engine)(nil)
 var _ hotstuff.Communicator = (*Engine)(nil)
 
 func NewEngine(
 	log zerolog.Logger,
 	net network.Network,
 	me module.Local,
-	prov network.Engine,
+	prov consensus.ProposalProvider,
 	core *Core,
 ) (*Engine, error) {
+
 	// Inbound FIFO queue for `messages.BlockResponse`s
+	// TODO can be removed along with https://github.com/dapperlabs/flow-go/issues/6254
```

Comment:
This whole saga with `messages.BlockResponse` started as workaround to mitigate timing issue between compliance engine and hotstuff, instead of processing `messages.BlockResponse` we could feed block proposals in loop as it was before.

Reply:
To add to Yurii's comment, `BlockResponse`s specifically come from the sync engine. On `master` we added logic to deal with those ranges of blocks sequentially (because of the race condition around HotStuff not marking a block as valid before its child is inserted to the state). The race condition isn't a problem with our refactoring for Active Pacemaker, so we don't need to deal with blocks from the sync engine differently from any other block we receive.